include/UtfView/null_term.hpp

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+#ifndef UTFVIEW_NULL_TERM_HPP
+#define UTFVIEW_NULL_TERM_HPP
+
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+namespace utfview {
+
+/* PAPER: namespace std { */
+
+/* PAPER */
+
+template <class I>
+concept exposition_only_default_initializable_and_equality_comparable_iter_value =
+    std::default_initializable<std::iter_value_t<I>> &&
+    std::equality_comparable_with<std::iter_reference_t<I>,
+                                  std::iter_value_t<I>>; // @*exposition only*@
+
+struct null_sentinel_t {
+  template <std::input_iterator I>
+    requires(not std::forward_iterator<I>) &&
+      exposition_only_default_initializable_and_equality_comparable_iter_value<I>
+  friend constexpr bool operator==(I const& it, null_sentinel_t) {
+    return *it == std::iter_value_t<I>{};
+  }
+  template <std::forward_iterator I>
```

Comment:
Why do we need both?  If you get rid of the forward overload, and get rid of the not-forward constraint on the input overload, the only thing you lose is passing an iterator by reference instead of by value.  Are the two here just for that small optimization?

Reply:
This is not a big thing; I'm essentially just wondering.

Reply:
Yeah, that's the only reason why. The paper has the following language about it:

> The `null_sentinel_t`'s `operator==` has a separate overload for input iterators that takes the iterator by reference instead of by value. We want to take input iterators by reference because they are not required to be copyable. However, for forward iterators, we want to take by value because otherwise we incur a double indirection (e.g. `int* const& it`) that compilers may not optimize.

Reply:
Gotcha.  Thanks.